slides/week4/week4.org

```diff
@@ -0,0 +1,87 @@
+#+END_NOTES
+- At a high level, your program has three kinds of memory
+ - *Static memory:* allocated at compile time for things like static or global variables
+ - *The stack:* managed automatically by the language
+  - A new stack frame is allocated when a function is called, and released when the function exits
+  - This is where your parameters and local variables are stored
+ - *The heap:* your space to do whatever you want
+  - You can dynamically ask for memory from the OS and return it, all at runtime
+  - You have to keep track of what you've allocated and release it when you're done
+- We're not going to go too in depth on any of these, but a high level understanding is useful
+- At an even higher level, all the memory in your computer is a huge, indexable, iterable space, divided into bytes
+
+* References
+#+BEGIN_NOTES
+Run through some basic examples here.
+#+END_NOTES
+- A reference is a variable that points to the same content as another variable
+- Why would you want that?
+ - Remember the stack? Normally, parameters are copied into the new stack frame so they can be manipulated without affecting the rest of the program
+ - But what if you /want/ a function to change some of its inputs? Use a reference!
+ - This can also improve performance, since you're only copying a memory address instead of a (potentially) larger object
+- You can't change where a reference points after initialization
+- You can mostly just treat references exactly like normal variables once they're initialized
+
+* Pointers
+#+BEGIN_NOTES
+Run through some basic examples here.
+#+END_NOTES
+- A pointer is a special type of variable that "points to" another variable
+- The actual content of a pointer variable is a memory address
+- How are these different from references?
+ - You have to explicitly dereference a pointer to access its target
+ - You can change where a pointer is pointing
+- Pointers are great for passing around heap objects, but how do you keep track of what objects are still in use?
+
+* Smart pointers
+#+BEGIN_NOTES
+Run through some basic examples here.
+#+END_NOTES
+- Wrap a normal pointer to make it easier to manage the lifecycle of objects
+- You should generally use one of these two types instead of a normal pointer:
+ - =shared_ptr=: multiple =shared_ptr= s can own a single object, and the object is destroyed when the last =shared_ptr= is destroyed.
+ - =unique_ptr=: only one =unique_ptr= can own an object, and the object is destroyed when the =unique_ptr= is destroyed.
+- You should initizialize these using the =make_shared= and =make_unique= functions, respectively
+
+* Casting, the wrong way
+- Sometimes you want to convert a variable to a different type: this is called casting
+- In C, you can simply change the type of a variable like so:
+ - ~int a = (int)not_an_int;~
+ - Now =a= is an =int= reference to =not_an_int=, which may or may not actually be an =int=
```

Comment:
_a_ is not a reference in this example. 
A better example for this might involve pointers since we've taught them that.
B *old_var = ...;
A *new_var = (A*)old_var;
Here, the compiler happily learns to treat whatever object is at the address in old_var as an object of type A. It will use the same bytes of data, but interpret them differently. This can be semantically meaningless depending on the specific types A and B.

For a more detailed example, which is too large for a slide, consider the following structs
struct A {
short member1;
};
struct B {
char member1;
char member2;
};
A *a = ...;
B *b = (B*)a;

A novice programmer may mistakenly think that the compiler will somehow know how to transform between these two types, not realizing that the compiler will implicitly just split the 2 bytes of a->member1 into b->member1 and b->member2. This will likely not have any semantic meaning in the domain of the application.
While this is a trivial example, as our classes become more complex, even more experienced programmers can mistakenly cast between incompatible types. the *_cast<>() functions discussed this week offer checks to ensure that various bad things like this don't happen.